### PR TITLE
ASoC: SOF: sof-pci-dev: add .shutdown callback

### PR DESCRIPTION
**Pull request by keyonjie**
Add .shutdown callback to handle reboot/poweroff related stuff, and pull
DSP back to D0I0 before reboot if it is at D0I3, to fix the boot fail
issue at the consequent rebooting.

Fixes: https://github.com/thesofproject/linux/issues/2571
Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/core.c
Comment (kv2019i) [inline sound/soc/sof/core.c:388]: PCI driver shutdown has "void" return type, so we can't return an error to upper layer. But, but, I think it's ok to keep the "int" and return type, and keep this aligned with rest of SOF ops API.

File: sound/soc/sof/intel/hda.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda.c]: Would be good to split the Intel specific part into a separate patch. I.e. one to add shutdown op, and another to add Intel HDA implementation of it.

The comment is a bit vague, maybe just leave it out and explain the change in commit.

File: sound/soc/sof/intel/hda.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda.c]: This seems sufficient as a comment.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @keyonjie . I think this looks ok. Please split to two patches (common/intel) and address the comments and then this is ready from my point of view.

**Comment by keyonjie** [comment]
Thanks, @kv2019i comments addressed now.

File: sound/soc/sof/core.c
Comment (lyakh) [inline sound/soc/sof/core.c]: can this just be
```
return snd_sof_shutdown(sdev);
```

File: sound/soc/sof/intel/hda.c
Comment (lyakh) [inline sound/soc/sof/intel/hda.c]: optional: you could `return 0;` here and remove `ret` initialisation at the top

**Review by lyakh — CHANGES_REQUESTED** [review]

**Comment by keyonjie** [comment]
Thanks @lyakh updated now.

File: sound/soc/sof/intel/hda.c
Comment (dbaluta) [reply sound/soc/sof/intel/hda.c]: Maybe a dumb question. But why DSP needs to be in D0 before shutting down? Are there any commands sent from Linux driver to DSP at this point and DSP needs to be up to process them?

Comment (kv2019i) [reply sound/soc/sof/intel/hda.c]: @dbaluta This is a hw/fw workaround effectively for Intel systems. With soft reset, some hw state is not reset by hw, we have existing systems where the state is not reset by BIOS, and it cannot be done from driver directly, so we end up with unable to bring the device up after the subsequent system boot after soft reset.

File: sound/soc/sof/intel/tgl.c
Comment (kv2019i) [inline sound/soc/sof/intel/tgl.c:26]: In commit, "Fixes:" is only used when pointing to commit hashes. To refer to a bug URL, please use "BugLink: " instead.

File: sound/soc/sof/ops.h
Comment (kv2019i) [inline sound/soc/sof/ops.h:37]: In commit: "platform specific shutdown stuff" -> "platform specific actions at shutdown"

**Review by kv2019i — CHANGES_REQUESTED** [review]
Two minor issues in git commit messages, otherwise good for me.

File: sound/soc/sof/intel/hda.c
Comment (dbaluta) [reply sound/soc/sof/intel/hda.c]: @kv2019i thanks! It makes sense.

File: sound/soc/sof/intel/tgl.c
Comment (dbaluta) [reply sound/soc/sof/intel/tgl.c:26]: @kv2019i since we are here and I know on i.MX there might be similar issues, do you know how the HIFI DSP is reset / power on / power off on Intel platforms?
Does the BIOS takes care of this. Or DSP by itself can initiate the rest/power off procedure?

Comment (kv2019i) [reply sound/soc/sof/intel/tgl.c:26]: @dbaluta Actual power-on is triggered via PCI MMIO writes, see hda-dsp.c:hda_dsp_core_power_down() in kernel. On DSP side, there is a sequence to ramp down resources and got a loop executing waiti, see for example platform_pm_runtime_power_off().

**Review by lyakh — DISMISSED** [review]
Thanks for fixing! Once comments from @kv2019i are addressed, no more suggestions from me

**Comment by keyonjie** [comment]
Thanks for reviewing @kv2019i @lyakh updated now.

**Review by kv2019i — DISMISSED** [review]
Thanks, looks good!

**Review by lgirdwood — COMMENTED** [review]
@keyonjie can you confirm that shutdown() wont re-load FW just to set the D0 state ?

**Comment by keyonjie** [comment]
> @keyonjie can you confirm that shutdown() wont re-load FW just to set the D0 state ?

yes, the shutdown() here only doing possible switching of the ADSP state, from D0I3 to D0, no FW reloading will happen.

**Comment by plbossart** [comment]
I looked into this before the Thanksgiving break and the .shutdown doesn't seem necessary.

See device_shutdown(), it does a put_device() at the end that will result in the PCI .remove() callback being called. I don't see a difference between the .shutdown and .remove here, so what does this PR bring?

Does this solve anything?

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: this doesn't seem very useful, the PCI core already resumes the device so it *IS* in DO

````c
static void pci_device_shutdown(struct device *dev)
{
    struct pci_dev *pci_dev = to_pci_dev(dev);
    struct pci_driver *drv = pci_dev->driver;

 

    pm_runtime_resume(dev);
````

**Comment by keyonjie** [comment]
> I looked into this before the Thanksgiving break and the .shutdown doesn't seem necessary.
> 
> See device_shutdown(), it does a put_device() at the end that will result in the PCI .remove() callback being called. I don't see a difference between the .shutdown and .remove here, so what does this PR bring?

Thanks for raising this, are you talking about the below put_device()? It is the symmetry one of the get_device at the beginning of device_shutdown(), so no PCI.remove() will be invoked? 

```
void device_shutdown(void)
{
	struct device *dev, *parent;

	wait_for_device_probe();
	device_block_probing();

	cpufreq_suspend();

	spin_lock(&devices_kset->list_lock);
	/*
	 * Walk the devices list backward, shutting down each in turn.
	 * Beware that device unplug events may also start pulling
	 * devices offline, even as the system is shutting down.
	 */
	while (!list_empty(&devices_kset->list)) {
		dev = list_entry(devices_kset->list.prev, struct device,
				kobj.entry);

		/*
		 * hold reference count of device's parent to
		 * prevent it from being freed because parent's
		 * lock is to be held
		 */
		parent = get_device(dev->parent);
		get_device(dev);
		/*
		 * Make sure the device is off the kset list, in the
		 * event that dev->*->shutdown() doesn't remove it.
		 */
		list_del_init(&dev->kobj.entry);
		spin_unlock(&devices_kset->list_lock);

		/* hold lock to avoid race with probe/release */
		if (parent)
			device_lock(parent);
		device_lock(dev);

		/* Don't allow any more runtime suspends */
		pm_runtime_get_noresume(dev);
		pm_runtime_barrier(dev);

		if (dev->class && dev->class->shutdown_pre) {
			if (initcall_debug)
				dev_info(dev, "shutdown_pre\n");
			dev->class->shutdown_pre(dev);
		}
		if (dev->bus && dev->bus->shutdown) {
			if (initcall_debug)
				dev_info(dev, "shutdown\n");
			dev->bus->shutdown(dev);
		} else if (dev->driver && dev->driver->shutdown) {
			if (initcall_debug)
				dev_info(dev, "shutdown\n");
			dev->driver->shutdown(dev);
		}

		device_unlock(dev);
		if (parent)
			device_unlock(parent);

		put_device(dev);
		put_device(parent);

		spin_lock(&devices_kset->list_lock);
	}
	spin_unlock(&devices_kset->list_lock);
}

```

> 
> Does this solve anything?

Yes, it solve https://github.com/thesofproject/linux/issues/2571 for us for most scenarios (but not 100% yet).

File: sound/soc/sof/intel/hda.c
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c]: > this doesn't seem very useful, the PCI core already resumes the device so it _IS_ in DO
> 
> ```c
> static void pci_device_shutdown(struct device *dev)
> {
>     struct pci_dev *pci_dev = to_pci_dev(dev);
>     struct pci_driver *drv = pci_dev->driver;
> 
>  
> 
>     pm_runtime_resume(dev);
> ```

We are not in runtime suspended but D0I3 when WoV is active, so no resume from PCI core neither. 

To be simple, when only WoV is active, DSP is put to D0I3, we need to switch it back to D0 before performing system reboot, otherwise, there might be reboot failure for DSP.

**Comment by plbossart** [comment]
> Yes, it solve #2571 for us for most scenarios (but not 100% yet).

well then, what does this solve? and what are the problem cases?

**Comment by kv2019i** [comment]
@plbossart wrote:
> well then, what does this solve? and what are the problem cases?

In short, if we reboot while DSP is running on the low-power clock, upon subsequent reboot, we can bring that hardware from reset. Thus this workaround to bring DSP to normal mode before shutdown.

This won't cover the (arguably low probability) case that a soft reboot is done for a system where kernel is unable to run the shutdowns properly and/or communication with DSP is not working. In this case,  a hard reset is required to recover device audio.

**Review by ranj063 — APPROVED** [review]

**Comment by plbossart** [comment]
Allow me to push back @kv2019i - I am having a hard-time understanding the problem statement.

when the system reboots, doesn't it close applications first? That would close the PCM device that was used for WoV and as a result transition to D0i0 - thus no longer running in low-power clock.

**Comment by kv2019i** [comment]
@plbossart wrote:
@> Allow me to push back @kv2019i - I am having a hard-time understanding the problem statement.
> 
> when the system reboots, doesn't it close applications first? That would close the PCM device that was used for WoV and as a result transition to D0i0 - thus no longer running in low-power clock.

I'll let @keyonjie to fill in. On sysv-init/systemd Linux, this would be true (unless systemd reboot is forced or sysrq-reboot, is used) services would be stopped before reboot, but apparently on Chrome this is not the case.

But yeah, I'll let @keyonjie comment. Originally the problematic clock was used outside WoV usage, so the problem would be hit on regular Linux desktops as well.

**Comment by plbossart** [comment]
> But yeah, I'll let @keyonjie comment. Originally the problematic clock was used outside WoV usage, so the problem would be hit on regular Linux desktops as well.

Right, if we are using the D0 transition to mean 'please use the faster clock' we remain dependent on what the firmware configuration might be. that's not robust by any stretch of imagination.

**Comment by keyonjie** [comment]
> Allow me to push back @kv2019i - I am having a hard-time understanding the problem statement.
> 
> when the system reboots, doesn't it close applications first? That would close the PCM device that was used for WoV and as a result transition to D0i0 - thus no longer running in low-power clock.

That's true, but since our D0->D0I3 transition happens at 5 seconds delayed work, there could be race there: e.g. the trigger stop just finish the transition to D0i0 at the exact point 5 seconds timeout, so the D0->D0I3 transition may happen following that...

My fix here actually can't eliminate the race thoroughly, that's the reason why we are still seeing corner case here, @kv2019i and @keqiaozhang .

I am now changing my mind to the thought I talked at the release meeting, let's do a thorough DSP powering off at .shutdown(), the validation is running at this point, let me change this PR to 'draft' first.

**Comment by keyonjie** [comment]
Hi @plbossart @kv2019i @ranj063 @lyakh please help to review the new implementation, basically I switch to call .remove() at .shutdown() for TGL platform, this will lead to longer reboot time (should be several ms level delta, not measure yet).

This behaves better during my local reboot stress test, it passes 30 iterations (the previous can't pass 10) and test is continue...

Hi @keqiaozhang can you help to test with the latest change at your end also.

**Comment by plbossart** [comment]
> > when the system reboots, doesn't it close applications first? That would close the PCM device that was used for WoV and as a result transition to D0i0 - thus no longer running in low-power clock.
> 
> That's true, but since our D0->D0I3 transition happens at 5 seconds delayed work, there could be race there: e.g. the trigger stop just finish the transition to D0i0 at the exact point 5 seconds timeout, so the D0->D0I3 transition may happen following that...

That means we are missing a cancel_work_sync() ?

**Comment by keyonjie** [comment]
> > > when the system reboots, doesn't it close applications first? That would close the PCM device that was used for WoV and as a result transition to D0i0 - thus no longer running in low-power clock.
> > 
> > 
> > That's true, but since our D0->D0I3 transition happens at 5 seconds delayed work, there could be race there: e.g. the trigger stop just finish the transition to D0i0 at the exact point 5 seconds timeout, so the D0->D0I3 transition may happen following that...
> 
> That means we are missing a cancel_work_sync() ?

Adding cancel_work_sync() can't eliminate the race thoroughly, I had another local change to add a mutex to protect this dsp_power_state access, but that can't make sure the delayed work won't be scheduled after trigger_stop (D0I3->D0), except we do a explicit cancel_work_sync() at PCM trigger stop. So with all this possibilities, adding an extra D0i3->D0 at .shutdown() looks the most light way to me.

On the other hand, to make things simpler and more robust, going with another way that just powering off the whole ADSP at .shutdown() (reflected in my latest update above) looks better to me, we can simplify with just power off DSP cores without doing other removal stuffs to speed up the .shutdown() running if needed. Please let me know what do you feel about that.

**Comment by plbossart** [comment]
> > > > when the system reboots, doesn't it close applications first? That would close the PCM device that was used for WoV and as a result transition to D0i0 - thus no longer running in low-power clock.
> > > 
> > > 
> > > That's true, but since our D0->D0I3 transition happens at 5 seconds delayed work, there could be race there: e.g. the trigger stop just finish the transition to D0i0 at the exact point 5 seconds timeout, so the D0->D0I3 transition may happen following that...
> > 
> > 
> > That means we are missing a cancel_work_sync() ?
> 
> Adding cancel_work_sync() can't eliminate the race thoroughly, I had another local change to add a mutex to protect this dsp_power_state access, but that can't make sure the delayed work won't be scheduled after trigger_stop (D0I3->D0), except we do a explicit cancel_work_sync() at PCM trigger stop. So with all this possibilities, adding an extra D0i3->D0 at .shutdown() looks the most light way to me.

I am lost here. If we have a design where we have a risk of race condition because of workqueue, then we need to fix this.
We've had problems before, we really need to make sure this is under control.

> On the other hand, to make things simpler and more robust, going with another way that just powering off the whole ADSP at .shutdown() (reflected in my latest update above) looks better to me, we can simplify with just power off DSP cores without doing other removal stuffs to speed up the .shutdown() running if needed. Please let me know what do you feel about that.

No objection if the race condition is handled.

**Comment by keyonjie** [comment]
> > Adding cancel_work_sync() can't eliminate the race thoroughly, I had another local change to add a mutex to protect this dsp_power_state access, but that can't make sure the delayed work won't be scheduled after trigger_stop (D0I3->D0), except we do a explicit cancel_work_sync() at PCM trigger stop. So with all this possibilities, adding an extra D0i3->D0 at .shutdown() looks the most light way to me.
> 
> I am lost here. If we have a design where we have a risk of race condition because of workqueue, then we need to fix this.
> We've had problems before, we really need to make sure this is under control.

Yes, let me submit a fix for that with another PR.

**Comment by keyonjie** [comment]
> > > Adding cancel_work_sync() can't eliminate the race thoroughly, I had another local change to add a mutex to protect this dsp_power_state access, but that can't make sure the delayed work won't be scheduled after trigger_stop (D0I3->D0), except we do a explicit cancel_work_sync() at PCM trigger stop. So with all this possibilities, adding an extra D0i3->D0 at .shutdown() looks the most light way to me.
> > 
> > 
> > I am lost here. If we have a design where we have a risk of race condition because of workqueue, then we need to fix this.
> > We've had problems before, we really need to make sure this is under control.
> 
> Yes, let me submit a fix for that with another PR.

Here it is: https://github.com/thesofproject/linux/pull/2600

File: sound/soc/sof/intel/tgl.c
Comment (ranj063) [inline sound/soc/sof/intel/tgl.c]: @keyonjie lets remove the comment. the code is self-explanatory

File: sound/soc/sof/ops.h
Comment (ranj063) [inline sound/soc/sof/ops.h:40]: does shutdown need to return int?

**Comment by ranj063** [comment]
@keyonjie I think the patches are good. Can you please remove the comment in the last patch and then the PR will be good to go. @sathya-nujella is testing it right now and once he reports back the results, we can merge this

File: sound/soc/sof/ops.h
Comment (keyonjie) [reply sound/soc/sof/ops.h:40]: > does shutdown need to return int?

the sof_pci_shutdown() itself doesn't require int return, here returning int with snd_sof_shutdown() is just to make it consistent with other snd_sof_xxx()s. This is exactly same with what we have for snd_sof_device_remove() and sof_pci_remove().

**Comment by keyonjie** [comment]
@ranj063 @kv2019i @bardliao updated.

File: sound/soc/sof/sof-priv.h
Comment (bardliao) [inline sound/soc/sof/sof-priv.h]: nit-pick: It seems the empty line is not required.

File: sound/soc/sof/intel/hda.h
Comment (bardliao) [inline sound/soc/sof/intel/hda.h]: Where does hda_dsp_shutdown be used?

**Review by bardliao — COMMENTED** [review]
@keyonjie I didn't see where is hda_dsp_shutdown() used. Do I miss something?

File: sound/soc/sof/intel/hda.h
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h]: > Where does hda_dsp_shutdown be used?

Good catch, this was previous used but now should be removed.

File: sound/soc/sof/sof-priv.h
Comment (keyonjie) [reply sound/soc/sof/sof-priv.h]: > nit-pick: It seems the empty line is not required.

Then I should change the comment for "probe and remove" to "probe/remove and shutdown"?

**Comment by keyonjie** [comment]
@bardliao updated now.

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/intel/tgl.c
Comment (cujomalainey) [reply sound/soc/sof/intel/tgl.c:26]: @kv2019i fixes is a github construct, it should be used in PR description to help autoclose related bugs if possible. But yes i agree not in the commit message.

**Review by ranj063 — APPROVED** [review]

**Pull request merged**
